Modules/_ssl.c

```diff
@@ -6241,6 +6338,41 @@ _ssl_get_default_verify_paths_impl(PyObject *module)
     return NULL;
 }
 
+/*[clinic input]
+@critical_section
+_ssl.get_sigalgs
+[clinic start generated code]*/
+
+static PyObject *
+_ssl_get_sigalgs_impl(PyObject *module)
+/*[clinic end generated code: output=ab0791b63856854b input=bf74cdad3a19d29e]*/
+{
+#if OPENSSL_VERSION_NUMBER >= 0x30400000L
+    const char *sigalgs;
+    PyObject *sigalgs_str, *sigalgs_list;
+
+    if ((sigalgs = SSL_get1_builtin_sigalgs(NULL)) == NULL) {
+        PyErr_NoMemory();
+        return NULL;
+    }
+
+    if ((sigalgs_str = PyUnicode_DecodeASCII(sigalgs, strlen(sigalgs),
+                                             "strict")) == NULL) {
+        OPENSSL_free((void *)sigalgs);
+        return NULL;
+    }
+
+    OPENSSL_free((void *)sigalgs);
+    sigalgs_list = PyUnicode_Split(sigalgs_str, _Py_LATIN1_CHR(':'), -1);
+    Py_DECREF(sigalgs_str);
+    return sigalgs_list;
+#else
+    PyErr_SetString(PyExc_NotImplementedError,
+                    "Getting signature algorithms requires OpenSSL 3.4 or later.");
+    return NULL;
```

Comment:
For a follow-up issue as well, we can have a helper function for this where we indicate the action we want to perform and the minimum version, e.g.:

```c
    raise_unsupported_operation("getting signature algorithms", 0x30400000L);
    return NULL;
```

I don't know if it's really necessary though but this may ensure that we consistently raise the same exception type.

Reply:
Yeah - I think it could make sense to use PyExc_NotImplementedError. I'd suggest passing a string version of the version, though, so the helper doesn't need to convert the integer form to a string. Since the helper won't actually be doing the comparison, there's no need for it to have access to the integer from of the version.

One slight complication is the AWS-LC issue we ran into. That wouldn't be able to use the helpful if the error string it creates refers to an OpenSSL version, unless we make callers pass in the complete error string and only have the helper take care of filling in the exception type.

Reply:
Ok.